src/link/MachO/commands.zig

```diff
@@ -0,0 +1,378 @@
+        };
+    }
+
+    pub fn deinit(self: *LoadCommand, allocator: *Allocator) void {
+        return switch (self.*) {
+            .Segment => |*x| x.deinit(allocator),
+            .Dylinker => |*x| x.deinit(allocator),
+            .Dylib => |*x| x.deinit(allocator),
+            .Unknown => |*x| x.deinit(allocator),
+            else => {},
+        };
+    }
+
+    fn writeStruct(command: anytype, writer: anytype) !void {
+        const slice = [1]@TypeOf(command){command};
+        return writer.writeAll(mem.sliceAsBytes(slice[0..1]));
+    }
+
+    fn eql(self: LoadCommand, other: LoadCommand) bool {
+        if (@as(@TagType(LoadCommand), self) != @as(@TagType(LoadCommand), other)) return false;
+        return switch (self) {
+            .DyldInfoOnly => |x| eqlStruct(x, other.DyldInfoOnly),
+            .Symtab => |x| eqlStruct(x, other.Symtab),
+            .Dysymtab => |x| eqlStruct(x, other.Dysymtab),
+            .Main => |x| eqlStruct(x, other.Main),
+            .VersionMin => |x| eqlStruct(x, other.VersionMin),
+            .SourceVersion => |x| eqlStruct(x, other.SourceVersion),
+            .LinkeditData => |x| eqlStruct(x, other.LinkeditData),
+            .Segment => |x| x.eql(other.Segment),
+            .Dylinker => |x| x.eql(other.Dylinker),
+            .Dylib => |x| x.eql(other.Dylib),
+            .Unknown => |x| x.eql(other.Unknown),
+        };
+    }
+
+    fn eqlStruct(lhs: anytype, rhs: anytype) bool {
+        return mem.eql(u8, mem.asBytes(&lhs), mem.asBytes(&rhs));
+    }
+};
+
+pub const SegmentCommand = struct {
+    inner: macho.segment_command_64,
+    sections: std.ArrayListUnmanaged(macho.section_64) = .{},
+
+    pub fn empty(inner: macho.segment_command_64) SegmentCommand {
+        return .{ .inner = inner };
+    }
+
+    pub fn read(alloc: *Allocator, reader: anytype) !SegmentCommand {
+        const inner = try reader.readStruct(macho.segment_command_64);
+        var segment = SegmentCommand{
+            .inner = inner,
+        };
+        try segment.sections.ensureCapacity(alloc, inner.nsects);
+
+        var i: usize = 0;
+        while (i < inner.nsects) : (i += 1) {
+            const section = try reader.readStruct(macho.section_64);
+            segment.sections.appendAssumeCapacity(section);
+        }
+
+        return segment;
+    }
+
+    pub fn write(self: SegmentCommand, writer: anytype) !void {
+        const cmd = [1]macho.segment_command_64{self.inner};
+        try writer.writeAll(mem.sliceAsBytes(cmd[0..1]));
+
+        for (self.sections.items) |sect| {
+            const section = [1]macho.section_64{sect};
+            try writer.writeAll(mem.sliceAsBytes(section[0..1]));
+        }
+    }
+
+    pub fn deinit(self: *SegmentCommand, alloc: *Allocator) void {
+        self.sections.deinit(alloc);
+    }
+
+    fn eql(self: SegmentCommand, other: SegmentCommand) bool {
+        if (!mem.eql(u8, mem.asBytes(&self.inner), mem.asBytes(&other.inner))) return false;
```

Comment:
This pattern is dangerous as the structure may contain hidden padding fields, if you want a field-by-field equality check there's `meta.eql`.

Ditto for some other occurrences of `mem.eql + mem.asBytes`.

Reply:
Is it though? `self.inner` is an `extern` struct so it has a guaranteed memory layout, and I've made sure I only compare byte representation that is valid. Otherwise, if what you are saying is true, I shouldn't trust `mem.asBytes` from you second comment either, right?

Reply:
> Is it though?

Yes, in the general case it is dangerous. Making sure the two types are `extern` is not enough, consider the following example:
```zig
const S = extern struct {
  a: u8,
  // _padding: [3]u8, // Inserted automatically to recover the ABI alignment for `b`
  b: u32,
};
var x = S{ .a = 1, .b = 42 };
var y = S{ .a = 1, .b = 42 };
```

If `x` and `y` are stack-allocated the padding area contains random data and thus using `mem.cmp` won't give you the answer you expect. This check is ok iff the structures are `packed` or `extern` but with no padding in between.
Same for the serialization part, you can trust `mem.asBytes` not to contain garbage if the condition above is true.

If all this structures do is wrapping the mach-o structs used for reading/writing data carry on then.

Reply:
Ah, that's a great point, thanks @LemonBoy! In this case though, as you've noticed, the structures are effectively glorified wrappers for Mach-O structs used exclusively for reading/writing the data they carry, and the `eql` method is there to ease with testing, that's all. Having said that, perhaps I should change the tests to instead directly compare to populated byte buffers instead having a custom `eql` method which can be confusing.

Reply:
On second thought, since I do want to test read and write ops separately, I've heeded your suggestion since it's bang on. In short, now relying on `meta.eql` where appropriate, thanks!